src/services/refactors/addOrRemoveBracesToArrowFunction.ts

```diff
@@ -0,0 +1,109 @@
+/* @internal */
+namespace ts.refactor.addOrRemoveBracesToArrowFunction {
+    const refactorName = "Add or remove braces in an arrow function";
+    const refactorDescription = Diagnostics.Add_or_remove_braces_in_an_arrow_function.message;
+    const addBracesActionName = "Add braces to arrow function";
+    const removeBracesActionName = "Remove braces from arrow function";
+    const addBracesActionDescription = Diagnostics.Add_braces_to_arrow_function.message;
+    const removeBracesActionDescription = Diagnostics.Remove_braces_from_arrow_function.message;
+    registerRefactor(refactorName, { getEditsForAction, getAvailableActions });
+
+    interface Info {
+        container: ArrowFunction;
+        expression: Expression;
+        addBraces: boolean;
+    }
+
+    function getAvailableActions(context: RefactorContext): ApplicableRefactorInfo[] | undefined {
+        const { file, startPosition } = context;
+        const info = getConvertibleArrowFunctionAtPosition(file, startPosition);
+        if (!info) return undefined;
+
+        return [{
+            name: refactorName,
+            description: refactorDescription,
+            actions: [
+                info.addBraces ?
+                    {
+                        name: addBracesActionName,
+                        description: addBracesActionDescription
+                    } : {
+                        name: removeBracesActionName,
+                        description: removeBracesActionDescription
+                    }
+            ]
+        }];
+    }
+
+    function getEditsForAction(context: RefactorContext, _actionName: string): RefactorEditInfo | undefined {
+        const { file, startPosition } = context;
+        const info = getConvertibleArrowFunctionAtPosition(file, startPosition);
+        if (!info) return undefined;
+
+        const { expression, container } = info;
+        const changeTracker = textChanges.ChangeTracker.fromContext(context);

```

Comment:
Both branches end up calling the same method on the change tracker. So this can be written as:

```ts
function getEditsForAction(context: RefactorContext, _actionName: string): RefactorEditInfo | undefined {
    const { file, startPosition } = context;
    const info = getConvertibleArrowFunctionAtPosition(file, startPosition);
    if (!info) return undefined;

    const { expression, container } = info;
    const newBody = _actionName === addBracesActionName ? createBlock([createReturn(expression)])
        :  _actionName === removeBracesActionName ? needsParentheses(expression) ? createParen(expression) : expression
        : Debug.fail("invalid action");
    const edits = textChanges.ChangeTracker.with(context, t => t.replaceNode(file, container, updateBody(container, newBody)));
    return { renameFilename: undefined, renameLocation: undefined, edits };
}

function updateBody(node: ArrowFunction, body: ConciseBody): ArrowFunction {
    return updateArrowFunction(node, node.modifiers, node.typeParameters, node.parameters, node.type, body);
}

function needsParentheses(expression: Expression): boolean {
    if (isLiteralExpression(expression)) {
        return false;
    }
    switch (expression.kind) {
        case SyntaxKind.Identifier:
        case SyntaxKind.NullKeyword:
        case SyntaxKind.TrueKeyword:
        case SyntaxKind.FalseKeyword:
        case SyntaxKind.UndefinedKeyword:
        case SyntaxKind.PropertyAccessExpression:
        case SyntaxKind.ElementAccessExpression:
        case SyntaxKind.CallExpression:
        case SyntaxKind.NewExpression:
            return false;
        default:
            return true;
    }
}
```

Reply:
Maybe `needsParentheses` should be written backwards -- what are the expressions that *do* need parentheses? `BinaryExpression` where the operator is `,`, `ObjectLiteralExpression`, any others?